[Hack Week] ApiFaker: introduce new module and logic

hichamboushaba opened the pull request:
### Description
As part of my Hack Week, I'm bringing back the ApiFaker project paqN3M-SI-p2, and I'm splitting the feature into multiple PRs.

This PR adds the `apifaker` gradle module, and with it the following changes:
1. A Room database to hold the list of faked endpoints.
2. The `EndpointProcessor` class is responsible for matching and finding the faked endpoints.
3. The `OkHttp` interceptor that replaces the response when needed. 

### Steps to reproduce
Code review and green CI should be enough.

But if you prefer to do some tests, you can use the following PR https://github.com/woocommerce/woocommerce-android/pull/13070

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Create a new module for the Api Faker

hichamboushaba committed:
Add Room DB for holding the endpoints

hichamboushaba committed:
Add initial logic for intercepting and faking responses

hichamboushaba committed:
Some changes to the DB schema

hichamboushaba committed:
Handle extracting data from the request for matching

hichamboushaba committed:
Add some logging, and handle errors in parsing

hichamboushaba committed:
Use hilt viewmodel

hichamboushaba committed:
Use an app's Fragment to allow theming the UI properly

hichamboushaba committed:
Start working on the details screen

hichamboushaba committed:
Handle remaining fields

hichamboushaba committed:
Handle saving changes

hichamboushaba committed:
Fixes to saving and navigation

hichamboushaba committed:
Rename classes

hichamboushaba committed:
Add some improvements to the DB schema and dao

hichamboushaba committed:
Improve the query statement

hichamboushaba committed:
Improve the query statement

hichamboushaba committed:
Minor improvements

hichamboushaba committed:
Add a way to enable/disable the API Faker

hichamboushaba committed:
Minor improvements to the home screen's UI

hichamboushaba committed:
Add an initial screen to list the different mocked endpoints

hichamboushaba committed:
Consume the status configuration

hichamboushaba committed:
Add ability to delete requests

hichamboushaba committed:
Add ability to delete requests

hichamboushaba committed:
Add HttpMethod to the DB entities

hichamboushaba committed:
Enable destructive migration

hichamboushaba committed:
Add support for filtering by http method

hichamboushaba committed:
Display a hint when ApiFaker is enabled

hichamboushaba committed:
Improve navigation

hichamboushaba committed:
Improve navigation

hichamboushaba committed:
UI improvements

hichamboushaba committed:
Apply internal visibility to some declarations

hichamboushaba committed:
Persist the ApiFaker status

hichamboushaba committed:
Add ability to disable the ApiFaker from the UI hint

hichamboushaba committed:
Add query parameters to the DB model

hichamboushaba committed:
Filter requests using query parameters

hichamboushaba committed:
Add a way to add query parameters in the UI

hichamboushaba committed:
Some fixes

hichamboushaba committed:
Fix endpoint extraction from Jetpack tunnel endpoints

hichamboushaba committed:
Log a warning message if multiple endpoints are matched

hichamboushaba committed:
Add an artifical delay

hichamboushaba committed:
Disable schema export

hichamboushaba committed:
Rename classes

dangermattic commented:
<!--
  0 Errors
  11 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          11 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ApiFakerConfig</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ApiFakerInterceptor</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ApiFakerUiHelper</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>ApiFakerDatabase</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>EndpointTypeConverter</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>QueryParameterConverter</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>Screen</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>EndpointDetailsViewModel</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>HomeViewModel</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/245">21.3</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Run unit tests of all libraries in CI

hichamboushaba committed:
Add the faker library to the app

hichamboushaba committed:
Add an entry to developer options and handle navigation to the API faker

hichamboushaba committed:
Suppress deprecation temporarily

hichamboushaba committed:
Fix detekt issue

hichamboushaba committed:
Match the TopAppBar background color to the app theme

hichamboushaba committed:
Fix detekt issue

hichamboushaba committed:
Bump db version

hichamboushaba committed:
Improve logic of handling Jetpack tunnel endpoints

This commit adds logic for extracting the method from the path, and also the logic to wrap the body using `data` property when needed

hichamboushaba committed:
Handle query parameters extraction for Jetpack tunnel requests

hichamboushaba committed:
Start config flow eagerly

Otherwise we won't get the preference value

hichamboushaba committed:
Fix issue with updating the faker config

hichamboushaba committed:
Set text color as white in the hint

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Fix detekt issues

hichamboushaba commented on libs/apifaker/src/main/java/com/woocommerce/android/apifaker/EndpointProcessor.kt:
```diff
@@ -0,0 +1,120 @@
+    private val jsonObjectProvider: JSONObjectProvider
+) {
+    fun fakeRequestIfNeeded(request: Request): Response? {
+        // TODO match against method and query parameters too
```
This TODO here is not accurate, as part of this PR we are already matching against the HTTP method, the comment is fixed in next PRs, to avoid issues with merge conflicts, let's ignore it here.

hichamboushaba committed:
Improve the logic of extracting the http method for Jetpack tunneled requests

hichamboushaba commented on libs/apifaker/src/main/java/com/woocommerce/android/apifaker/EndpointProcessor.kt:
```diff
@@ -0,0 +1,122 @@
+            val (path, method, body) = if (method == "GET") {
+                Triple(
+                    url.queryParameter("path")!!.substringBefore("&"),
+                    url.queryParameter("_method") ?: "GET",
```
For Jetpack Tunneled GET requests, the HTTP method is sent as a query parameter `_method`, see https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/64479b5648444c945dccae776f596daf192f4b87/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/jetpacktunnel/JetpackTunnelGsonRequest.kt#L249

This will generally be always `GET`, but to avoid hardcoding it, I'm reading it from the parameter when it exists.

hichamboushaba commented on libs/apifaker/src/main/java/com/woocommerce/android/apifaker/EndpointProcessor.kt:
```diff
@@ -0,0 +1,122 @@
+                val jsonObject = jsonObjectProvider.parseString(originalBody)
+                val pathParts = jsonObject.getString("path").split("&")
+                val path = pathParts[0]
+                val method = pathParts.firstOrNull { it.startsWith("_method") }?.split("=")?.getOrNull(1) ?: "POST"
```
For Jetpack Tunneled POST requests, the actual endpoint HTTP method will be past with the json property `path`, see: https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/64479b5648444c945dccae776f596daf192f4b87/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/jetpacktunnel/JetpackTunnelGsonRequest.kt#L265

hichamboushaba commented on libs/apifaker/src/main/java/com/woocommerce/android/apifaker/EndpointProcessor.kt:
```diff
@@ -0,0 +1,122 @@
+            url.encodedPath.trimEnd('/').matches(Regex(JETPACK_TUNNEL_REGEX)) &&
+            !startsWith("{\"data\":")
+        ) {
+            "{\"data\": $this}"
```
Jetpack Tunnel sends responses [wrapped](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/64479b5648444c945dccae776f596daf192f4b87/fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/jetpacktunnel/JetpackTunnelResponse.kt#L4) in a `data` property, this logic makes it easier for faking those without the wrap, which should work for both site credentials login and Jetpack login.

JorgeMucientes replied:
This brings up the same question from above. What if we don't want to have JSON formatted body with `data { }` as the root of if?  Like the example mentioned above when simulating an error that returns html?

hichamboushaba replied:
As said in the other thread, I would love to keep this as is right now, changing it would require more testing.

Depending on the developer's needs, we can revisit this in the future.

hichamboushaba committed:
Mark ApiFaker log tag as internal

hichamboushaba commented on .buildkite/commands/run-unit-tests.sh:
```diff
@@ -2,7 +2,7 @@
 echo "--- 🧪 Testing"
 set +e
-./gradlew testJalapenoDebugUnitTest lib:cardreader:testDebugUnitTest lib:iap:testDebugUnitTest
+./gradlew testJalapenoDebugUnitTest testDebugUnitTest
```
This allows running unit tests for all modules instead of having to list all of them

hichamboushaba committed:
Remove explicit argument type

Zulu-1540 reviewed (changes requested):
Fake account

Zulu-1540 reviewed (dismissed):
Hacker

hichamboushaba committed:
Add some padding between the text fields and their captions

hichamboushaba committed:
Add more tests

JorgeMucientes commented on libs/apifaker/build.gradle:
```diff
@@ -0,0 +1,72 @@
+    maven {
+        url 'https://a8c-libs.s3.amazonaws.com/android'
+        content {
+            includeGroup "org.wordpress"
+            includeGroup "org.wordpress.fluxc"
+            includeGroup "org.wordpress.fluxc.plugins"
+            includeGroup "org.wordpress.wellsql"
+            includeGroup "org.wordpress.mediapicker"
+            includeGroup "com.automattic"
+            includeGroup "com.automattic.tracks"
```
Do you need all these? I see you only add FluxC as dependency (which also makes sense since it's where all the requests happen). But why include `tracks`, `mediapicker`, `wellsql`, `wordpress` ?

hichamboushaba replied:
I don't like this either, but I tried to just keep the same setup as what we have in the other modules, personally I think this code should be in the root build.gradle file, or better in a [convention plugin](https://docs.gradle.org/current/samples/sample_convention_plugins.html), but since this is not the setup we have yet, I prefer to keep the same setup as the other modules here, as the `includeGroup` statements won't have any impact here, they are just used for defining the repo itself.

JorgeMucientes commented on libs/apifaker/src/main/java/com/woocommerce/android/apifaker/ApiFakerInterceptor.kt:
```diff
@@ -0,0 +1,39 @@
+                .protocol(Protocol.HTTP_1_1)
+                .message("Fake Response")
+                .code(fakeResponse.statusCode)
+                // TODO check if it's safe to always use JSON as the content type
```
Since the body of the faked request is anything the user inputs from the app, wouldn't it make more sense to leave `MediaType` null? For example sometimes we get and html body for API errors. Haven't tested yet, but what happens if we keep this `application/json` content type and enter non json body?

hichamboushaba replied:
This is a good point; but for now I'll keep it like this, my rationale for hardcording `application/json` here is that all of the API requests we have in the app use JSON responses, the HTML errors are an exception that can be caused by a specific site setup that returns an error before the site handles the error.

> what happens if we keep this application/json content type and enter non json body?

It should work as you expect from the app, the upper layer will try to deserialize the content, and would fail, then return a JSON error, this is what happens when we receive HTML instead of JSON (for example this case: p1723565504566789/1720009283.216679-slack-C6H8C3G23 `API org.json.JSONException: Value <!DOCTYPE of type java.lang.String cannot be converted to JSONObject`)

JorgeMucientes reviewed (approved):
Great job @hichamboushaba, looking forward to see this in action in subsequent PRs. Code looks clean and well organised. I just left a couple minor questions, but nothing blocking.

hichamboushaba committed:
Make DB flow collection lifecycle aware

hichamboushaba committed:
Fix UI with long endpoint paths

hichamboushaba committed:
Improve logic of warning log

This makes sure the log is not printed when the matched endpoints use different query parameters.

hichamboushaba committed:
Merge pull request #13083 from woocommerce/task/api-faker-5

[Hack Week] ApiFaker: match endpoints using query parameters

hichamboushaba committed:
Merge pull request #13076 from woocommerce/task/api-faker-4

[Hack Week] ApiFaker: display a hint on the main app UI when the ApiFaker is enabled

hichamboushaba committed:
Merge pull request #13070 from woocommerce/task/api-faker-3

[Hack Week] ApiFaker: Adding, deleting and deleting endpoints

hichamboushaba committed:
Merge pull request #13064 from woocommerce/task/api-faker-2

[Hack Week] ApiFaker: Home screen and integration with the app

hichamboushaba committed:
Merge branch 'trunk' into task/api-faker-1

# Conflicts:
#	WooCommerce/src/main/res/values/strings.xml

hichamboushaba committed:
Add Compose icons dependency to the apiFaker lib
